src/python/pants/backend/codegen/tasks/simple_codegen_task.py

```diff
@@ -78,6 +103,31 @@ def codegen_targets(self):
     """
     return self.context.targets(self.is_gentarget)
 
+  @classmethod
+  def _codegen_strategy_for_name(cls, name):
+    return {
+      'global': SimpleCodegenTask.GlobalCodegenStrategy,
+      'isolated': SimpleCodegenTask.IsolatedCodegenStrategy,
+    }[name]
+
+  @property
+  def codegen_strategy(self):
+    """Returns the codegen strategy object used by this codegen.
+
+    This is controlled first by the forced_codegen_strategy property, then by user-specified
+    options if the former returns None.
+
+    If you just want the name ('global' or 'isolated') of the strategy, use codegen_strategy.name.
+
+    :return SimpleCodegenTask.CodegenStrategy: the codegen strategy object.
+    """
+    strategy = self.forced_codegen_strategy
+    if strategy is None:
+      strategy = self.get_options().strategy
+    if self._codegen_strategy is None or self._codegen_strategy.name != strategy:
```

Comment:
To me, the "double" invocation below `self._codegen_strategy_for_name(strategy)(self)` is awkward and I'm trying to think of a way to make it easier to follow.   Essentially, its just a factory method but returns a function instance which is kind of unusual.

1) You could  move the instantiation into `_codegen_strategy_for_name()` 
2)  Dispense with the dictionary lookup. It looks like it is just a way to make python look like it supports a case statement.
3) If `_codegen_strategy_for_name()` was an instance method, you wouldn't have to explicityly pass `self`

Then  you could write something with one or two fewer cases of indirection like:

```
def _codegen_strategy_for_name(self, strategy):
  if strategy == 'global':
    return SimpleCodegen.GlobalCodegenStrategy(self)
  return SimpleCodegen.IsolatedCodegenStrategy(self) 
```

Then its more clearly following a factory  method pattern and  clearer (to me) that the instance variable makes sense.

3) You mentioned that it is used for testing.  I don't see why you need to use the factory for the test.   If you were to just instantiate the strategy you want and you'd end up using the same amount of code in your test.

```
global_strategy = SimpleCodegen.GlobalCodegenStrategy(self)
isolated_strategy = SimpleCodegen.IsolatedCodegenStrategy(None)

...
global_dirs = set(global_strategy.codegen_workdir_suffix(target) for target in targets)
isolated_dirs = set(isolated_strategy.codegen_workdir_suffix(target) for target in targets)
```